File: src/main/java/lan/dk/podcastserver/business/WatchListBusiness.java
package lan.dk.podcastserver.business;

import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.WatchList;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.repository.WatchListRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Set;
import java.util.UUID;

/**
 * Created by kevin on 17/01/2016 for PodcastServer
 */
@Component
public class WatchListBusiness {

    private final WatchListRepository watchListRepository;
    private final ItemRepository itemRepository;

    @Autowired
    public WatchListBusiness(WatchListRepository watchListRepository, ItemRepository itemRepository) {
        this.watchListRepository = watchListRepository;
        this.itemRepository = itemRepository;
    }

    public WatchList findOne(UUID id) {
        return watchListRepository.findOne(id);
    }

    public List<WatchList> findAll() {
        return watchListRepository.findAll();
    }

    public Set<WatchList> findContainsItem(UUID itemId) {
        Item item = itemRepository.findOne(itemId);
        return watchListRepository.findContainsItem(item);
    }

    public WatchList add(UUID watchListId, UUID itemId) {
        WatchList watchList = watchListRepository.findOne(watchListId);
        Item item = itemRepository.findOne(itemId);

        return watchListRepository.save(watchList.add(item));
    }

    public WatchList remove(UUID watchListId, UUID itemId) {
        WatchList watchList = watchListRepository.findOne(watchListId);
        Item item = itemRepository.findOne(itemId);

        return watchListRepository.save(watchList.remove(item));
    }

    public void delete(UUID uuid) {
        watchListRepository.delete(uuid);
    }

    public WatchList save(WatchList watchList) {
        return watchListRepository.save(watchList);
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/business/WatchListBusiness.kt
package com.github.davinkevin.podcastserver.business

import arrow.core.getOrElse
import com.github.davinkevin.podcastserver.entity.WatchList
import com.github.davinkevin.podcastserver.service.JdomService
import com.github.davinkevin.podcastserver.utils.k
import lan.dk.podcastserver.repository.ItemRepository
import lan.dk.podcastserver.repository.WatchListRepository
import org.springframework.stereotype.Component
import java.util.*

/**
 * Created by kevin on 17/01/2016 for PodcastServer
 */
@Component
class WatchListBusiness(val watchListRepository: WatchListRepository, val itemRepository: ItemRepository, val jdomService: JdomService) {

    fun add(watchListId: UUID, itemId: UUID): WatchList {
        val watchList = watchListRepository.findById(watchListId).k()
                .getOrElse { throw RuntimeException("Watchlist not found") }
        val item = itemRepository.findById(itemId).k()
                .getOrElse { throw RuntimeException("Item with ID $itemId not found") }

        return watchListRepository.save(watchList.apply { add(item) })
    }

    fun remove(watchListId: UUID, itemId: UUID): WatchList {
        val watchList = watchListRepository.findById(watchListId).k()
                .getOrElse { throw RuntimeException("Watchlist not found") }
        val item = itemRepository.findById(itemId).k()
                .getOrElse { throw RuntimeException("Item with ID $itemId not found") }

        return watchListRepository.save(watchList.apply { remove(item) })
    }

    fun delete(uuid: UUID) = watchListRepository.deleteById(uuid)

    fun asRss(id: UUID, domainFromRequest: String) = watchListRepository.findById(id).k()
            .map { jdomService.watchListToXml(it, domainFromRequest) }
            .getOrElse { throw RuntimeException("Rss generation of watchlist $id caused Error") }
}

